### PR TITLE
Register Cupyx

### PR DESCRIPTION
**Pull request by quasiben**
Fixes https://github.com/dask/distributed/issues/3547

PR adds cupyx to list of registered (de-)serializers

This does pass but it does produce some nastiness at the end (we can probably ignore but thought it was worth mentioning)

```
distributed/protocol/tests/test_cupy.py::test_serialize_cupy_sparse[cuda-dtype0-coo_matrix] PASSED
distributed/protocol/tests/test_cupy.py::test_serialize_cupy_sparse[cuda-dtype0-csc_matrix] PASSED
distributed/protocol/tests/test_cupy.py::test_serialize_cupy_sparse[cuda-dtype0-csr_matrix] malloc_consolidate(): invalid chunk size
Fatal Python error: Aborted

Current thread 0x00007f6989001740 (most recent call first):
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/logging.py", line 628 in _runtest_for_main
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/contextlib.py", line 81 in __enter__
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/logging.py", line 607 in _runtest_for
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/contextlib.py", line 81 in __enter__
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/logging.py", line 645 in pytest_runtest_setup
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/callers.py", line 182 in _multicall
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/manager.py", line 81 in <lambda>
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/manager.py", line 87 in _hookexec
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/hooks.py", line 289 in __call__
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/runner.py", line 210 in <lambda>
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/runner.py", line 237 in from_call
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/runner.py", line 210 in call_runtest_hook
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/runner.py", line 185 in call_and_report
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/runner.py", line 93 in runtestprotocol
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/runner.py", line 84 in pytest_runtest_protocol
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/manager.py", line 81 in <lambda>
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/manager.py", line 87 in _hookexec
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/hooks.py", line 289 in __call__
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/main.py", line 271 in pytest_runtestloop
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/manager.py", line 81 in <lambda>
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/manager.py", line 87 in _hookexec
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/hooks.py", line 289 in __call__
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/main.py", line 247 in _main
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/main.py", line 197 in wrap_session
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/main.py", line 240 in pytest_cmdline_main
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/manager.py", line 81 in <lambda>
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/manager.py", line 87 in _hookexec
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pluggy/hooks.py", line 289 in __call__
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/_pytest/config/__init__.py", line 93 in main
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/site-packages/pytest/__main__.py", line 7 in <module>
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/runpy.py", line 85 in _run_code
  File "/datasets/bzaitlen/miniconda3/envs/20200302/lib/python3.6/runpy.py", line 193 in _run_module_as_main
```

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thanks.  Is that `Fatal Python error: Aborted` crashing the test runner, or an individual worker thread, or something else?

**Comment by quasiben** [comment]
Upon second look I think there are other issues.  i'll report back

**Comment by jakirkham** [comment]
Maybe try running with `MALLOC_CHECK_=1` based on [this SO answer]( https://stackoverflow.com/a/3718867/3877089 )?

It's interesting that this comes up in the `"cuda"` serialization case and not the `"dask"` one.

**Comment by quasiben** [comment]
There are issues with the `dia_matrix`.  Working on the commit now

**Comment by jakirkham** [comment]
Ah ok. Thanks for looking at this Ben. Happy to help if I can 🙂

**Comment by jakirkham** [comment]
Was wondering if maybe Cython was involved in this bug. However it looks like CuPy is defining all of these sparse matrices (including base classes) in pure Python. So that doesn't seem like it.

**Comment by quasiben** [comment]
It only happens on the next iteration of a test

**Comment by jakirkham** [comment]
Yeah my guess is pytest magic is somehow interfering in the normal cleanup process of some CuPy objects. We saw [something similar with RMM]( https://github.com/rapidsai/rmm/pull/275#discussion_r377986498 ). A little mystified as to exactly what is causing it in this case though.

**Comment by quasiben** [comment]
I am splitting up the tests to move on.  And also amending the dtypes as we were seeing this error:

```
        if dtype != 'f' and dtype != 'd' and dtype != 'F' and dtype != 'D':
            raise ValueError(
>               'Only float32, float64, complex64 and complex128'
                ' are supported')
E           ValueError: Only float32, float64, complex64 and complex128 are supported
```

**Comment by jakirkham** [comment]
Though the dtypes already met these requirements as they were all 32-bit and 64-bit. Not sure I understand the error message 😕

Was testing endianness to make sure both big and little endian worked for end users.

**Comment by quasiben** [comment]
Even moving all tests in separate functions the errors still arises.  i've skipped them for now and would like to return to it when i have more time

File: distributed/protocol/tests/test_cupy.py [hunk 1]
```diff
@@ -66,20 +67,88 @@ def test_serialize_cupy_from_rmm(size):
...
+    "dtype", [numpy.dtype("<f4"), numpy.dtype("<f8"),],
+)
+@pytest.mark.parametrize("serializer", ["cuda", "dask",])
+def test_serialize_cupy_sparse_csc(sparse_type, dtype, serializer):
+    a_host = numpy.array([[0, 1, 0], [2, 0, 3], [0, 4, 0]], dtype=dtype)
+    a = cupy.asarray(a_host)
+
+    anz = a.nonzero()
+    acoo = cupy_sparse.coo_matrix((a[anz], anz))
+    asp = sparse_type(acoo)
```
Comment (jakirkham) [inline distributed/protocol/tests/test_cupy.py]: Following up from our offline discussion, I wonder if we can drop this `coo_matrix` part and just construct the sparse matrix directly. Sounds like the intermediate step of building a `coo_matrix` is running into some edge case (Ben's observation).

**Comment by quasiben** [comment]
Here is a minimal repro which sometime (though not always) generates a segfault

```python
import numpy
import cupy
import cupyx.scipy.sparse as cupy_sparse

from distributed.protocol import deserialize, serialize

dtype = numpy.dtype("<f8")
data = cupy.array([0, 1, 2, 3], dtype)
indices = cupy.array([0, 1, 3, 2], 'i')
indptr = cupy.array([0, 2, 3, 4], 'i')

asp = cupy_sparse.csr_matrix((data, indices, indptr), shape=(3, 4))
header, frames = serialize(asp, serializers=["cuda"])
asp2 = deserialize(header, frames)

print(asp.todense())
cupy.testing.assert_allclose(asp.todense(), asp2.todense())
```

**Comment by quasiben** [comment]
During a time where we succeed and running with `nvprof` we see an uneven number of free/Malloc calls

```
      API calls:   63.03%  859.26ms         1  859.26ms  859.26ms  859.26ms  cudaFree
                   33.07%  450.88ms        13  34.683ms  10.420us  450.68ms  cudaMalloc
                    1.80%  24.502ms      1528  16.035us     117ns  2.0051ms  cuDeviceGetAttribute
                    1.41%  19.218ms        16  1.2011ms  1.1621ms  1.3129ms  cuDeviceTotalMem
                    0.18%  2.4662ms        10  246.62us  200.39us  342.74us  cuModuleLoadData
                    0.14%  1.9325ms        16  120.78us  111.08us  141.94us  cuDeviceGetName
```

**Comment by quasiben** [comment]
My think now is that there is an issue with using `register_generic` with sparse cupy arrays.  We do patch some behavior for native cupy arrays and I'm wondering if we shouldn't just be doing the same for sparse data as well

**Comment by jakirkham** [comment]
Is the `print` needed to reproduce?

**Comment by jakirkham** [comment]
So this is interesting...

```python
In [1]: import numpy 
   ...: import cupy 
   ...: import cupyx.scipy.sparse as cupy_sparse 
   ...:  
   ...: from distributed.protocol import deserialize, serialize 
   ...:  
   ...: dtype = numpy.dtype("<f8") 
   ...: data = cupy.array([0, 1, 2, 3], dtype) 
   ...: indices = cupy.array([0, 1, 3, 2], 'i') 
   ...: indptr = cupy.array([0, 2, 3, 4], 'i') 
   ...:  
   ...: asp = cupy_sparse.csr_matrix((data, indices, indptr), shape=(3, 4)) 
   ...: header, frames = serialize(asp, serializers=["cuda"]) 
   ...: asp2 = deserialize(header, frames) 
   ...:  
   ...: print(asp.toarray()) 
   ...: cupy.testing.assert_allclose(asp.toarray(), asp2.toarray())                                                                                    
[[0. 1. 0. 0.]
 [0. 0. 0. 2.]
 [0. 0. 3. 0.]]
Segmentation fault (core dumped)
```

Also it always segfaults for me.

**Comment by jakirkham** [comment]
Tried preallocating the dense CuPy arrays to write into to see if that helped. Appears that will still run into issues. So there may be some memory management issues in the `todense`/`toarray` functions.

```python
import numpy
import cupy
import cupyx.scipy.sparse as cupy_sparse

from distributed.protocol import deserialize, serialize

dtype = numpy.dtype("<f8")
data = cupy.array([0, 1, 2, 3], dtype)
indices = cupy.array([0, 1, 3, 2], 'i')
indptr = cupy.array([0, 2, 3, 4], 'i')

asp = cupy_sparse.csr_matrix((data, indices, indptr), shape=(3, 4))
header, frames = serialize(asp, serializers=["cuda"])
asp2 = deserialize(header, frames)

asp_dense = cupy.empty(shape=asp.shape, dtype=asp.dtype, order="F")
asp2_dense = cupy.empty(shape=asp2.shape, dtype=asp2.dtype, order="F")

asp.toarray(order="F", out=asp_dense)
asp2.toarray(order="F", out=asp2_dense)

cupy.testing.assert_allclose(asp_dense, asp2_dense)
```

Note: The F-ordering was required `csr_matrix` had some issues with C-order.

**Comment by jakirkham** [comment]
Ah! I think I figured it out. We weren't initializing the RMM pool. 😅

**Comment by quasiben** [comment]
Should we need to initialize the RMM pool ? We don't do this for cupy arrays, do we ?

**Comment by quasiben** [comment]
After chatting with John offline I have a better understanding of `register_generic` and it _should_ work.  As an experiment, I explicitly defined the (de-)serialization needed for all sparse matrix types in cupy and things pass now.  The constructors for cupy sparse object due vary -- perhaps this variation lead to some complications in the generic case ?

**Comment by jakirkham** [comment]
Turns out the segfaulting was due to a pointer in `MatDescriptor` getting passed through. Have fixed the serialization behavior of `MatDescriptor` in PR ( https://github.com/dask/distributed/pull/3556 ). Also pulled in the registration fix from here (thanks Ben! 😄). Finally did a little test cleanup, which should make things easier to follow (in addition to fixing the `dia_matrix` test failure).

**Comment by quasiben** [comment]
closing in favor of #3556

**Comment by jakirkham** [comment]
Thanks for debugging this with me Ben! 😄

**Pull request closed without merging**